Feat: ng-[event]-prepare class for structural animations (ngIf etc)

Narretz opened the pull request:
A fix / workaround for #12969

We currently don't have any place in the docs that documents the classes ngAnimate adds. At least I can't find it .:o I think it got lost during 1.4. We should add it back so people can find this feature easier.

Idea by @matsko

Narretz commented:
@petebacondarwin This is a feature that is needed as a workaround for an ngAnimate issue that can't currently be solved in another way. Should we try to get it in before 1.5.0?

petebacondarwin commented:
Can we include links in the commit messages to the issues and other commits that this is fixing or related to?

If it a fix to a bug then it is fine to put it in 1.5.0.

It is not clear from the tests how this solve the issue in #12969 - we should have some test or at least a description of the issue and how this PR fixes that in the code and/or commit message. (Or am I just missing that?)

Narretz commented:
It's not a direct fix, more of a workaround.
Basically, you provide a style (e.g. opacity:0) for ng-enter-prepare that makes sure the element is hidden before the animation actually starts. We cannot directly test this, because the flicker does not always appear.

Narretz commented:
@petebacondarwin I've added an example to the issue https://github.com/angular/angular.js/issues/12969#issuecomment-160612044

petebacondarwin commented:
OK, so we also need docs for this feature, especially explaining how it can be used to fix problems that might appear like #12969.

Narretz commented:
I'll add some docs for it tomorrow

Narretz commented:
@petebacondarwin I've added docs, fancy taking a look?

petebacondarwin commented:
Will do. Later today

matsko commented:
LGTM

Narretz committed:
feat(ngAnimate): provide ng-[event]-prepare class for structural animations

The new prepare class is added before the animation is pushed to the
queue and removed before the animation runs, i.e. it is immediately
available when a structural animation (enter, leave, move)
is initialized.

The class can be used to apply CSS to explicitly hide these elements
to prevent a flash of content before the animation runs.
This can happen if a structural animation (such as ng-if) sits at the
bottom of a tree which has ng-class animations on the parents.
Because child animations are spaced out with requestAnimationFrame,
the ng-enter class might not be applied in time, so the ng.if element is
briefly visible before its animation starts.

Narretz commented:
@matsko I've noticed that the prepare classes aren't applied to $animateCss based animations. I guess this expected, as they run outside of the animationQueue?
